modules/entity_version_history/src/Controller/EntityVersionHistoryController.php

```diff
@@ -0,0 +1,180 @@
+    $header = [
+      $this->t('Version'),
+      $this->t('Title'),
+      $this->t('Date'),
+      $this->t('Created by'),
+    ];
+
+    $rows = [];
+
+    $build['entity_version_history_table'] = [
+      '#theme' => 'table',
+      '#rows' => $rows,
+      '#header' => $header,
+    ];
+
+    $build['pager'] = ['#type' => 'pager'];
+
+    return $build;
+  }
+
+  /**
+   * Checks access to the history route.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The account accessing the route.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The actual route match of the route.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   The access result.
+   */
+  public function checkAccess(AccountInterface $account, RouteMatchInterface $route_match): AccessResultInterface {
+    $entity = $this->getEntityFromRouteMatch($route_match);
+    $cache = new CacheableMetadata();
+    $cache->addCacheContexts(['route']);
+
+    if (!$entity) {
+      return AccessResult::forbidden('No entity found in the route.')->addCacheableDependency($cache);
+    }
+
+    $cache->addCacheableDependency($entity);
+    $bundle = $entity->bundle();
+    $history_storage = $this->entityTypeManager->getStorage('entity_version_history_settings');
+    $cache->addCacheTags($history_storage->getEntityType()->getListCacheTags());
+
+    if (!$config_entity = $history_storage->load($entity->getEntityTypeId() . '.' . $bundle)) {
+      return AccessResult::forbidden('No history settings found for this entity type and bundle.')->addCacheableDependency($cache);
+    }
+
+    $cache->addCacheableDependency($config_entity);
+    $cache->addCacheContexts(['user.permissions']);
+
+    if (!$account->hasPermission('access entity version history')) {
+      return AccessResult::forbidden('Insufficient permissions to access the entity version history page.')->addCacheableDependency($cache);
+    }
+
+    return AccessResult::allowed()->addCacheableDependency($cache);
+  }
+
+  /**
+   * Provides a title callback for the history page.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The route match.
+   *
+   * @return string
+   *   The title for the history page.
+   */
+  public function title(RouteMatchInterface $route_match): string {
+    if ($entity = $this->getEntityFromRouteMatch($route_match)) {
+      return sprintf('History for %s', $entity->label());
+    }
+
+    return 'History';
+  }
+
+  /**
+   * Returns the current entity from a given route match.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The route match.
+   *
+   * @return \Drupal\Core\Entity\ContentEntityInterface|null
+   *   The entity or NULL if none exists.
+   */
+  protected function getEntityFromRouteMatch(RouteMatchInterface $route_match): ?ContentEntityInterface {
+    $route = $route_match->getRouteObject();
+    if (!$route || !$route->getOption('parameters')) {
```

Comment:
Why is this conditional needed?

Route will always be there and you don't need to check for the "paremeters" option as the route has the getParameters() method..